### ISSUE TITLE
remove efficiency and add spare clues

### ISSUE THREAD

**Issue by Zamiell** [Zamiell/hanabi-live#2210]
@timotree3

**Comment by jack67889** [comment]
Plea to keep efficiency. For end-game efficiency is useless and spare clues could be nice, but as an overall metric for how the game is going and what kind of clues you could afford, efficiency gives more info than spare clues.

Furthermore, I think spare clues would be nice (instead of number finesses required) in the dropdown. But it’s often misleading to rely on for tricky end-games.  First of all, tricky end-games are usually more about who can draw what. Secondly, you still need to count if you can actually use all the spare clues. I would say that count is just as much effort as counting what to do in the first place.

Recent game example. Spare clues would be 4 in this scenario. We need only 3 clues to stall. But still we are 1 clue short if Markus plays instead of discard.
![end-game](https://user-images.githubusercontent.com/74204920/124458821-bd3dc000-dd8d-11eb-9b15-ef030c47e145.png)

**Comment by Zamiell** [comment]
> as an overall metric for how the game is going and what kind of clues you could afford, efficiency gives more info than spare clues

clarify this statement because it seems wrong. do you know what spare clues are

**Comment by jack67889** [comment]
Yes I do (remaining possible clues - cards not gotten). But if you know spare clues is 2. I would say that is less info to decide if you can afford a 1-for-1 (mid-game) than when you know (future required) efficiency is 0.85. (And for end-game it’s the other way around.)

**Comment by Zamiell** [comment]
thats not what spare clues are

**Comment by jack67889** [comment]
I mean, that are quite literally clues to spare, but please enlighten me.

Or are you just being pedantic and saying I should also subtract clues to stall exactly 1 round (player count - 1)? If so: stalling exactly 1 round is pretty random, I don't see why you want to know that at the start of every game. Anyway, subtracting a constant doesn't change anything about what I wrote earlier. (For the example spare clues would be 1 in that case and still you are a clue short if you play.)

**Comment by timotree3** [comment]
I think the spare clues=4 metric is helpful in that endgame. Spare clues assumes you save all the 4's for the final round, so you if you play both 4's you'll only have 2 clues, which is not enough, therefore you must not play both 4's.

**Comment by timotree3** [comment]
Would it be satisfactory to show remaining possible clues and cards not gotten in the efficiency section and let players do the subtraction or division?

**Comment by Zamiell** [comment]
i dont think so, it is an important design goal that players should do no mental arithmetic

**Comment by Zamiell** [comment]
to clarify my statement - you could keep future eff and stick it in the tooltip

**Comment by jack67889** [comment]
I definitely agree it's quite a useful metric, like I said, but imho not better than (future required) efficiency. During a game you usually want to know if you should discard or give a (inefficient) clue, for which I think efficiency provides better info.

A lot of games are lost on end-game, for which spare clues would be more useful than efficiency. But I tried to demonstrate it's not such an important metric that it should replace efficiency. If you can figure out if the assumptions hold, you basically already solved the problem I would say. (I think knowing the assumptions is usually less important for situations where you use efficiency.)

And I like it to be a single metric, also for remaining possible clues the same assumptions apply of course.

**Comment by vEnhance** [comment]
To make sure I understand correctly, this is currently implemented as "number finesses required"?
If so, I suggest we can close this.

**Comment by DarthGandalf** [comment]
@vEnhance no.
A big difference is that efficiency is shown always, while "number finesses required" is shown only when you hover mouse on some obscure number somewhere.

**Comment by vEnhance** [comment]
Oh I meant whether they were the same number, not the UI presentation.

I will say I aesthetically prefer having future eff shown and spare clues in tooltip rather than vice-versa; the former number feels more intuitive to me. I understand this is a point of debate though.

**Comment by jack67889** [comment]
Spare clues = remaining possible clues - cards not gotten
Number finesses required = cards not gotten - remaining possible clues

So no, but spare clues is more useful imo. Especially since _number finesses required_ has no direct relation with finesses. So I would replace that, but I wouldn't swap places with efficiency (see arguments above).

**Comment by Zamiell** [comment]
> I aesthetically prefer having future eff shown
a single positive integer shown seems like the most intuitive solution and is the one that i currently prefer

**Comment by jack67889** [comment]
Spare clues and future required efficiency are based on the same numbers. The first one is the absolute value, the second one relative. The absolute value only really matters at the end of the game, but says nothing about the game state. If you want to make a decision based on the game state (can I afford to spend an extra clue) you want to know the relative value, since the same absolute value can map to very good or bad relative game states. So if you only know the absolute value, that contains less relevant info for most of the game.

**Comment by vEnhance** [comment]
So sounds like we want to have both spare clues and future required efficiency, and the question is just which one should be visible always and which one is hidden within the tool-tip. I agree with jack's previous post, meaning I vote for future efficiency to be the visible one. Would be curious if anyone else wants to vote.

Also, let `x := remaining possible clues - cards not gotten`. The website could print either max(x,0) or just x itself for "spare clues". Former is more informative, latter is more intuitive, maybe. I vote former here.

**Comment by jack67889** [comment]
I don't know, a negative number doesn't seem equivalent to 0 to me, so maybe don't show a number at all in that case. But a negative number seems perfectly fine to me; consider that for harder variants it will be negative for most of the time.

**Comment by Zamiell** [comment]
> The first one is the absolute value, the second one relative. The absolute value only really matters at the end of the game, but says nothing about the game state.

i don't understand this formulation.
if spare clues are at 2, and then you give 1-for-1s for the rest of the game, then spare clues will finish at 2.
if spare clues are at 2, and then you give a 0-for-1, then spare clues will drop to 1.

thus, it seems dynamical and relative in nature, communicating the same information regardless of whether you are in the middle or at the end of the game
can't you reformulate

**Comment by jack67889** [comment]
Tried to write in more general terms the examples I gave before.

But let's say spare clues is 2 at the start of the game and future required efficiency is 0.94. Once there are 8 cards left to get and spare clues is still/again 2, future required efficiency will be 0.8.

**Comment by Zamiell** [comment]
so you are saying that information is lost because having spare clues at the end of the game is slightly more important than having spare clues in the middle of the game

**Comment by Zamiell** [comment]
i think the counterargument is that the dynamical nature could make it harder to plan for the future.
the current paradigm is that players make binary strategy decisions based on current future eff and an arbitrary threshold. (i use somewhere in the range of 1.0 to 1.2, flopping my decision if it exceeds the threshold).

this is problematic because i cannot simply use the existing shown number as my threshold.
what i actually have to do is play out the line and examine future eff at the end of the line.
in the world of spare clues, i might still have to do this, but the big difference is that subsequent 0 for 1s in the line are just simple subtractions to the spare clues, which is easy enough to understand and compute mentally.
namely, i dont have to actually rely on the interface + doing a double hypothetical for every single turn in which i want to make a binary decision.

overall, i accept that some information loss is happening
in other words, players would have to mentally take into account that "spare clues are worth more in the end game".
the consequence being that the same decision based on a spare clue threshold would be "incorrect" in the mid game and "correct" in the end game.
but that might be acceptable if we can get the other benefits described in my previous paragraph.

**Comment by jack67889** [comment]
The absolute value of clues available only matters when you want to spend them all. The relative value incorporates the game state and is more useful as a stand-alone measure of clue usage. 

Making a nice profit is more important for your business than the exact amount of money in the bank, unless your bills are due.

**Comment by jack67889** [comment]
> i find this a surprising example to use to argue for future eff, since it displays 0.00, which essentially communicates "you've won" and provides no value beyond that.
> when contrasted to spare clues, at least you are getting some useful information there, even if it can be potentially misleading in the way you described above

The intention was to show that I agree spare clues is valuable, but certainly not as valuable as efficiency. Spare clues matters in the end-game and there efficiency is of course useless. But from my experience with tricky end-games, play/draw order is often more important than clue usage.

**Comment by jack67889** [comment]
By the way, I hope it was useful, but I don't feel like I added much to the discussion anymore. So I rather just try something if you're not convinced at this point, if there is no more input.

**Comment by vEnhance** [comment]
I don't suppose we have the UI space to have both? We could make the clue log a bit smaller and have an extra line...

**Comment by Zamiell** [comment]
I don't want to have both, because you mostly need one or the other

**Comment by vEnhance** [comment]
If we did choose spare clues over efficiency, where would we put the variant-required-efficiency?

**Comment by Romain672** [comment]
Just run into that subject.
What you want to know is until which player you can delay until the end game start.
Here is an old picture I shared on discard in depth: https://cdn.discordapp.com/attachments/776176933120507964/993775313869885492/startendgame.jpg
You take the first player (which is displayed in the top right), you advance by one player for each number on that picture, and you get one player X.
Then you move back one spot for each bomb.

Then, to know which player can start the end game, every time you want for the next player to start the end game, you need to keep one non-5 played during the end game.


This can be done automatiquely since it's globally known information.

Per example, in the picture shared in the first post, Markus started the game. It's 4players/5stuits, associated in my picture to number 2. So jeff is the player X.
There is one bomb, so you move back one spot, you arrive on Helana.
So you could decide to add a special icon on Helana.
Then, we want jeff to start the end game, that mean we want one non-5 to be played during the end game: b4 is impossible, and r4 is needed. So Markus should keep r4.



And I believe having that icon would be better. But it's babysitting some players, one more step into playing instead of them.
It would be a nice tool to have if it's done automatiquely and I would give a yes on having it.

**Comment by Romain672** [comment]
Note it's totally possible to do both of them. So even if you goes for the proposition on that subject, I could totally open a new Issue to add this icon.

**Comment by Zamiell** [comment]
> where would we put the variant-required-efficiency?

in the tooltip

**Comment by Romain672** [comment]
Btw, here is the formula to get the player which 'will maybe start last turn':
p+(k-p-1)*2+s

k is the number of card in the deck (k=50/45/44/40 for 2p/3p/4p/5p for 6 suits games)
k-1 is the number of card in the deck which will be used before the last turn.
p (=24) assume all no-5 have been played and correspond to the turn where a no-5 is played. Since it's not always the case, there is an extra step at the end for when it's not.
k-1-p is the number of discard + number of 5 played. Since you gain one clue out of them, i multiply them by 2.
s (=8) is the number of starting clue.

Then, take the modulo of number of players.
You call the first player 'player X' ;
then, for each bomb, you step back the player X by one slot.
then, for each missing non-5 card, you advance the player X by one slot.

**Comment by Romain672** [comment]
Here is another example: https://hanab.live/replay/895614#51
Bob is holding g4+g5, but there is another g4 on deck which could be drawn by someone else.
Having the "spare clue: X" will not help much to make Bob understand if he should clue a 5 or play.
With my method, you would see that Timehoodie would be the player X.
Which mean you would want either zero or four non-5 to be played during last turn. Four is impossible, so we are going for zero. Which mean playing g4 has no drawback.

**Comment by Romain672** [comment]
So all that to said that the argument that 'spare clue' is better for end game should not be taken into account imo.

And it could be habbit, but I think I would prefer effeciency+what i said to be visible.
